packages/flutter_tools/test/general.shard/flutter_tester_device_test.dart

```diff
@@ -0,0 +1,258 @@
+            'example.dill'
+          ],
+          stdout: 'Observatory listening on http://localhost:1234',
+          stderr: 'failure',
+          exitCode: 0,
+        )
+      ]);
+    });
+
+    testUsingContext('skips setting observatory port and uses the input port for for DDS instead', () async {
+      final TestFlutterTesterDevice testDevice = TestFlutterTesterDevice(enableObservatory: true);
+      await testDevice.start(compiledEntrypointPath: 'example.dill');
+      await testDevice.observatoryUri;
+
+      final Uri uri = await testDevice.ddsServiceUriFuture();
+      expect(uri.port, 1234);
+    }, overrides: <Type, Generator>{
+      ProcessManager: () => fakeProcessManager,
+    });
+  });
+}
+
+/// A Flutter Tester device.
+///
+/// Uses a mock HttpServer. We don't want to bind random ports in our CI hosts.
+class TestFlutterTesterDevice extends FlutterTesterTestDevice {
+  TestFlutterTesterDevice({
+    bool enableObservatory = false,
+    List<String> dartEntrypointArgs = const <String>[],
+  }) : super(
+    id: 999,
+    shellPath: '/',
+    debuggingOptions: DebuggingOptions.enabled(
+      const BuildInfo(
+        BuildMode.debug,
+        '',
+        treeShakeIcons: false,
+        packagesPath: '.dart_tool/package_config.json',
+      ),
+      startPaused: false,
+      disableDds: false,
+      disableServiceAuthCodes: false,
+      hostVmServicePort: 1234,
+      nullAssertions: false,
+      dartEntrypointArgs: dartEntrypointArgs,
+    ),
+    enableObservatory: enableObservatory,
+    machine: false,
+    host: InternetAddress.loopbackIPv6,
+    buildTestAssets: false,
+    flutterProject: null,
+    icudtlPath: null,
+    compileExpression: null,
+    fontConfigManager: FontConfigManager(),
+  );
+
+  final Completer<Uri> _ddsServiceUriCompleter = Completer<Uri>();
+
+  Future<Uri> ddsServiceUriFuture() => _ddsServiceUriCompleter.future;
+
+  @override
+  Future<DartDevelopmentService> startDds(Uri uri) async {
+    _ddsServiceUriCompleter.complete(uri);
+    final MockDartDevelopmentService mock = MockDartDevelopmentService();
+    when(mock.uri).thenReturn(Uri.parse('http://localhost:${debuggingOptions.hostVmServicePort}'));
+    return mock;
+  }
+
+  @override
+  Future<HttpServer> bind(InternetAddress host, int port) async => MockHttpServer();
+
+  @override
+  Future<StreamChannel<String>> get remoteChannel async => StreamChannelController<String>().foreign;
+}
+
+class MockDartDevelopmentService extends Mock implements DartDevelopmentService {}
+
+class MockHttpServer extends Mock implements HttpServer {}
+
+class MockProcessManager extends Mock implements ProcessManager {}
+
+class MockProcess extends Mock implements Process {
+  @override
+  Future<int> get exitCode async => 0;
+
+  @override
+  Stream<List<int>> get stdout => const Stream<List<int>>.empty();
+
+  @override
+  Stream<List<int>> get stderr => const Stream<List<int>>.empty();
+}
+
+class MockPlatform extends Mock implements Platform {}
```

Comment:
FakePlatform should be used instead of MockPlatform generally, should be just as easy to use.

Reply:
Looks like this was done in #75757, updated the branch to match. I left the `MockProcess` as it is, instead of using `FakeProcess`, as what we have in-place today is more suitable for performing assertions on the `environment` passed. IIUC `FakeProcess` will stall and eventually timeout the test when the environment doesn't match, which is harder to debug when something goes wrong as compared to what we have currently.